### ISSUE TITLE
Client just stops receiving stuff after ~30 seconds

### ISSUE THREAD

**Issue by rotolonico** [off99555/Unity3D-Python-Communication#14]
I am working on a screen capture using a bridge between Python and Unity with this library.

This is my client code:

```
public class HelloRequester : RunAbleThread
{
    /// <summary>
    ///     Request Hello message to server and receive message back. Do it 10 times.
    ///     Stop requesting when Running=false.
    /// </summary>
    protected override void Run()
    {
        ForceDotNet.Force(); // this line is needed to prevent unity freeze after one use, not sure why yet
        using (RequestSocket client = new RequestSocket())
        {
            client.Connect("tcp://localhost:6789");

            for (int i = 0; i < 100000 && Running; i++)
            {
                client.SendFrame("Hello");
                // ReceiveFrameString() blocks the thread until you receive the string, but TryReceiveFrameString()
                // do not block the thread, you can try commenting one and see what the other does, try to reason why
                // unity freezes when you use ReceiveFrameString() and play and stop the scene without running the server
//                string message = client.ReceiveFrameString();
//                Debug.Log("Received: " + message);
                byte[] bytes = null;
                bool gotMessage = false;
                while (Running)
                {
                    gotMessage = client.TryReceiveFrameBytes(out bytes); // this returns true if it's successful
                    if (gotMessage) break;
                }

                if (gotMessage)
                {
                    var colors = new List<Color>();
                    for (var p = 0; p < bytes.Length; p += 4)
                        colors.Add(new Color(bytes[p + 2] / 255f, bytes[p + 1] / 255f, bytes[p] / 255f));

                    HelloClient.pixels = colors;
                    HelloClient.update = true;
                }
            }
        }

        NetMQConfig.Cleanup(); // this line is needed to prevent unity freeze after one use, not sure why yet
    }
}
```

This is my server code:
```
import time
import zmq
from mss import mss

context = zmq.Context()
socket = context.socket(zmq.REP)
socket.bind("tcp://*:6789")

while True:
    #  Wait for next request from client
    message = socket.recv()

    #  Do some 'work'.
    #  Try reducing sleep time to 0.01 to see how blazingly fast it communicates
    #  In the real world usage, you just need to replace time.sleep() with
    #  whatever work you want python to do, maybe a machine learning task?
    time.sleep(0.1)

    #  Send reply back to client
    #  In the real world usage, after you finish your work, send your output here
    with mss() as sct:
        # Use the 1st monitor
        monitor = sct.monitors[0]

        # Grab the picture
        im = sct.grab(monitor)

        # Get the entire PNG raw bytes
        socket.send(im.raw)
```

(I tried to keep it as similar as possible as the hello world example)

If I set the time.sleep() to 2 seconds, I can get exactly 14 requests through until the client just stops getting any messages from the python server (it just gets stuck in the while gotMessage loop with no errors). If I set the time.sleep() to lower I can get more requests through but it seems like the connection just stops working after around ~30 seconds. What am I doing wrong?

Thanks!

**Comment by offchan42** [comment]
If it gets stuck in the loop then it means that it's not getting any message from the server.
Please print im.raw before socket.send() to see that it's continuously being sent.
I suspect that your sct.grab() code blocks after 30 seconds is passed.

**Comment by rotolonico** [comment]
> If it gets stuck in the loop then it means that it's not getting any message from the server.
> Please print im.raw before socket.send() to see that it's continuously being sent.
> I suspect that your sct.grab() code blocks after 30 seconds is passed.

Hey, thanks for the reply; I think I figured out what was happening. Both client and servers were in a mutual deadlock and were waiting for a response from one another. Implementing the [lazy pirate pattern](https://zguide.zeromq.org/docs/chapter4/#Client-Side-Reliability-Lazy-Pirate-Pattern) I got from this [Stack Overflow response](https://stackoverflow.com/questions/41009900/python-zmq-operation-cannot-be-accomplished-in-current-state) solved my problem.
